src/client/auth/gssapi.rs

```diff
@@ -0,0 +1,341 @@
+    // is established, an optional final token that must be sent to the server
+    // may be returned; otherwise another token to pass to the server is
+    // returned and the client context remains in the pending state.
+    async fn step(&mut self, challenge: &[u8]) -> Result<Option<Vec<u8>>> {
+        if challenge.is_empty() {
+            Err(Error::authentication_error(
+                GSSAPI_STR,
+                "Expected challenge data for GSSAPI continuation",
+            ))
+        } else {
+            if let Some(pending_ctx) = self.pending_ctx.take() {
+                match pending_ctx.step(challenge).map_err(|e| {
+                    Error::authentication_error(GSSAPI_STR, &format!("GSSAPI step failed: {}", e))
+                })? {
+                    Step::Finished((ctx, token)) => {
+                        self.is_complete = true;
+                        self.established_ctx = Some(ctx);
+                        Ok(token.map(|t| t.to_vec()))
+                    }
+                    Step::Continue((ctx, token)) => {
+                        self.pending_ctx = Some(ctx);
+                        Ok(Some(token.to_vec()))
+                    }
+                }
+            } else {
+                Err(Error::authentication_error(
+                    GSSAPI_STR,
+                    "Authentication context not initialized",
+                ))
+            }
+        }
+    }
+
+    // Perform the final step of Kerberos authentication by gss_unwrap-ing the
+    // final server challenge, then wrapping the protocol bytes + user principal.
+    // The resulting token must be sent to the server.
+    fn do_unwrap_wrap(&mut self, payload: &[u8]) -> Result<Vec<u8>> {
+        if let Some(mut established_ctx) = self.established_ctx.take() {
+            let _ = established_ctx.unwrap(payload).map_err(|e| {
+                Error::authentication_error(GSSAPI_STR, &format!("GSSAPI unwrap failed: {}", e))
+            })?;
+
+            if let Some(user_principal) = self.user_principal.take() {
+                let bytes: &[u8] = &[0x1, 0x0, 0x0, 0x0];
+                let bytes = [bytes, user_principal.as_bytes()].concat();
+                let output_token = established_ctx.wrap(false, bytes.as_slice()).map_err(|e| {
+                    Error::authentication_error(GSSAPI_STR, &format!("GSSAPI wrap failed: {}", e))
+                })?;
+                Ok(output_token.to_vec())
+            } else {
+                Err(Error::authentication_error(
+                    GSSAPI_STR,
+                    "User principal not specified",
+                ))
+            }
+        } else {
+            Err(Error::authentication_error(
+                GSSAPI_STR,
+                "Authentication context not established",
+            ))
+        }
+    }
+
+    fn is_complete(&self) -> bool {
+        self.is_complete
+    }
+}
+
+fn canonicalize_hostname(hostname: &str, mode: &CanonicalizeHostName) -> Result<String> {
```

Comment:
I wrote some unit tests for this but ultimately removed them since they are trivial. The more interesting testing of this will be indirect via the end-to-end GSSAPI auth tests in the next ticket.

Reply:
Yea, I'm personally trying to move away from unit tests that just explicitly encode basic code.